[Translation] added support for adding custom message formatter

aitboudad opened the pull request:
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | #6009, #10152, one item in #11742, #11948 |
| License | MIT |
| Doc PR | ~ |

aitboudad commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -214,7 +223,19 @@ public function transChoice($id, $number, array $parameters = array(), $domain =
+    {
+        @trigger_error('The '.__METHOD__.' method is deprecated since version 3.1 and will be removed in 4.0. Use trans instead.', E_USER_DEPRECATED);
+
+        $parameters['__number__'] = $number;
```
One possible BC break: we need to preserve `number` argument for the transChoice

Nicofuma commented:
What will be the replacement for `$translator->transChoice('some_message2', 10, array('%count%' => 10));` ?

`$translator->trans('some_message2', ['__number__' => 10, '%count%' => 10], null, null, TransChoiceMessageFormatter::NAME);` ?

It looks a bit weird to me. I think that if trans and transChoice are merged together we should have a way to automatically choose between the 2 (maybe automatically takes a '%count%' key is their isn't any '__number__' one or something like that?)

aitboudad commented:
@Nicofuma merging them together looks better!

iltar commented:
How would you envision this in the templates? Adding class names/constants is pretty annoying in twig. It also means that I have to pass _a lot_ of default values if I want to specify that. It feels like you have a lot more boilerplate to achieve the same. Having to write down `__number__` doesn't feel ideal either but having to rely on another pre-defined key is kind of magic as well.

What I know from experiences:
 - For transchoice, you always have to define a number
 - specifying the domain and locale is kind of annoying for most cases for the solution you present (messages domain/default locale)
 - A lot of times I translate without parameters but with a different domain

So I think we can agree that something is bothersome in the current implementation but i don't think the solution you present will solve it and will possibly even decrease the DX.

aitboudad commented:
Using multiple formatter is something rare but this allow us to migrate safely to [IntlMessageFormatter](https://github.com/symfony/symfony/pull/15068#issuecomment-117635934).
Right now if we combine trans with transChoice in one formatter it'll reduce some boilerplate,

`__number__` is provider for only BC break not for end-users, we can use `%count%` like we did in [twig transChoice](https://github.com/symfony/symfony/blob/master/src/Symfony/Bridge/Twig/Extension/TranslationExtension.php#L98)
```twig
{{ 'Hello world'|transchoice(count) }} //Before
{{ 'Hello world'|trans({ '%count%': count }) }} // After
```

iltar commented:
So how would that be determined? I can have a `%count%` in my arguments without it being a transchoice.

```php
$translator->trans($v, ['%count%' => 10], null, null, TransChoiceMessageFormatter::NAME);
// vs
$translator->transChoice($v, 10);
```

So where would it be less boilerplate? Maybe I missed something?

aitboudad commented:
@iltar 
> So how would that be determined? I can have a %count% in my arguments without it being a transchoice.

agreed we should find another pre-defined key :/

> Where would it be less boilerplate? Maybe I missed something?

As I said before TransChoiceMessageFormatter will be meged into TransMessageFormatter so we don't need to pass the formatter argument by default but we need to rely on a pre-defined key

HeahDude commented:
What about `trans_count` ?

iltar commented:
Ah okay, so it's the part where the key defines what type is used. In php you could use a constant but that would be cumbersome in twig. You could expose the value like `{{ 'foo' | trans({trans.count: 10}) }}`.

Something else that popped my mind, simply no key: `$translator->trans('foo', [10])`. Sadly it's typehinted as array for this case. In twig it could be `{{ 'foo' | trans(10) }}` though.

javiereguiluz commented:
At first glance this looks to me like a serious step back in DX:

```twig
{# before #}
{{ 'Hello world'|transchoice(count) }}

{# after #}
{{ 'Hello world'|trans({ '%count%': count }) }}
```

```php
// before
$translator->transChoice('some_message2', 10, array('%count%' => 10));

// after
$translator->trans('some_message2', [TransMessageFormatter::TRANS_CHOICE => 10]);
```

My questions: What's wrong with the current `transChoice()` method? What's the problem we're trying to solve? Why the new code is longer to type and harder to learn? Thanks!

mvrhov commented:
@javiereguiluz: take a look at how [message formatter](http://php.net/manual/en/messageformatter.formatmessage.php) class works

javiereguiluz commented:
@mvrhov yes, but I'd prefer if we first focus on the before/after changes for end users and then we care about the internal implementation. For now I feel that this proposal is worse than the current situation.

aitboudad commented:
@javiereguiluz having 2 fonctions that achieve the same behavior looks wrong to me 
and I think `tranchoice` will be useless for custom formattter such as [IntlMessageFormatter](http://php.net/manual/en/class.messageformatter.php) or [ExpressionsMessageFormatter](https://github.com/symfony/symfony/issues/17610). The main goal for me is to make IntlMessageFormatter as default formatter.

we can keep `tranchoice` as a helper method but IMO it shouldn't be part of [TranslatorInterface](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Translation/TranslatorInterface.php#L48)
if you have any suggestions I would be happy to implement it ;)

iltar commented:
@aitboudad Actually the only times I've used the transchoice, I've done so via twig and never in the code. I find it a rare occasion having to use the translator directly in the code in the first place (but does happen).

javiereguiluz commented:
@aitboudad here it is how Translator works today from the controller and the templates:

```php
// simple
$this->get('translator')->trans('Hello');
// placeholders
$this->get('translator')->trans('Hello %name%', ['%name%' => 'Fabien']);
// pluralization
$this->get('translator')->transChoice(
    'There is one apple|There are %count% apples',
    10,
    array('%count%' => 10)
);
// selecting the catalogue
$this->get('translator')->trans('Hello', [], 'admin');
// selecting the locale
$this->get('translator')->trans('Hello', [], 'messages', 'fr_FR');
```

```twig
{# simple #}
{% trans %}Hello{% endtrans %}
{{ 'Hello'|trans }}

{# placeholders #}
{% trans with { '%name%': 'Fabien' } %}Hello %name%{% endtrans %}
{{ 'Hello'|trans({'%name%': 'Fabien'}) }}

{# pluralization #}
{% transchoice count %}
    {0} There are no apples|{1} There is one apple|]1,Inf[ There are %count% apples
{% endtranschoice %}

{{ '{0} There ...|{1} There ...'|transchoice(count) }}

{# selecting the catalogue #}
{% trans from 'admin' %}Hello{% endtrans %}
{{ 'Hello'|trans({}, 'app') }}

{# selecting the locale #}
{% trans into 'fr_FR' %}Hello{% endtrans %}
{{ 'Hello'|trans({}, 'messages', 'fr_FR') }}
```

Please, provide the equivalent examples for the new way of using the Translator that you are proposing, so we can compare. Thanks!

aitboudad commented:
@javiereguiluz 
For php:
```patch
 // pluralization
-$this->get('translator')->transChoice(
+$this->get('translator')->trans(
     'There is one apple|There are %count% apples',
-    10,
-    array('%count%' => 10)
+    array(TransMessageFormatter::TRANS_CHOICE => 10)
 );
```
in Twig (I defined global variable transchoice = TransMessageFormatter::TRANS_CHOICE):
```patch
 {# pluralization #}
-{% transchoice count %}
+{% trans with { transchoice: 'Fabien' } %}
     {0} There are no apples|{1} There is one apple|]1,Inf[ There are %count% apples
-{% endtranschoice %}
+{% endtrans %}
 
-{{ '{0} There ...|{1} There ...'|transchoice(count) }}
+{{ '{0} There ...|{1} There ...'|trans({transchoice: count}) }}
```

javiereguiluz commented:
@aitboudad thanks for the info. I've updated your comment to add the missing `use` import.

Does the new system mean that `%count%` is a hardcoded variable name in translation messages? If not, how could we implement this example with the new proposal?

```php
$this->get('translator')->trans(
    'There is one apple|There are %num% apples',
    5,
    array('%num%' => 'five')
);
```

aitboudad commented:
@javiereguiluz 
```php
$this->get('translator')->trans(
    'There is one apple|There are %num% apples',
    array(TransMessageFormatter::TRANS_CHOICE => 5, '%num%' => 'five')
);
```
we can style use `%count%`:
```php
$this->get('translator')->trans(
    'There is one apple|There are %count% apples',
    array(TransMessageFormatter::TRANS_CHOICE => 5, '%count%' => 'five')
);
```

sstok commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -438,4 +471,17 @@ private function getConfigCacheFactory()
+        $name = $name ?: $this->defaultFormatter;
+
+        foreach ($this->formatters as $formatter) {
+            if ($name === $formatter->getName()) {
```
Can't you store the name as array key when registering?
This loop will be executed for **every** trans() call!

aitboudad replied:
fixed thanks!

aitboudad commented:
I've updated the PR and I think it's is ready for review,

Please note I added an optional formatter `PluralMessageFormatterInterface` for those who want to still use `transChoice` for pluralization and leave `transChoice` as it is.

stof commented:
for Twig, your diff is wrong. this is how it would look like:

```diff
 {# pluralization #}
-{% transchoice count %}
+{% trans with { (transchoice): count } %}
     {0} There are no apples|{1} There is one apple|]1,Inf[ There are %count% apples
-{% endtranschoice %}
+{% endtrans %}

-{{ '{0} There ...|{1} There ...'|transchoice(count) }}
+{{ '{0} There ...|{1} There ...'|trans({(transchoice): count}) }}
```

Without the braces, the key will not be an expression and so won't read the variable.

aitboudad commented:
@stof thanks I decided to leave deprecation `transChoice` for another PR or even for `3.3`, let us to discuss about the formatter part right now :)

stof commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml:
```diff
@@ -28,6 +28,12 @@
+
+        <service id="translator.formatter" alias="translator.formatter.default" />
+
+        <service id="translator.formatter.default" class="Symfony\Component\Translation\Formatter\TransMessageFormatter" public="false">
```
I would not name the service ``.default`` but ``.trans_message``, to match the name of the class (and make it easier to change the default later, if we don't have to rename services for that, as the actual default is handled by the alias, not by the service name)

stof commented on src/Symfony/Component/Translation/Formatter/MessageFormatterInterface.php:
```diff
@@ -0,0 +1,37 @@
+     *
+     * @return string
+     */
+    public function format($id, $locale, array $parameters = array());
```
should not be the id IMO, but the string retrieved from the catalogue (so that formatters don't have to be aware of the catalogue)

aitboudad replied:
you mean renaming the `$id` --> `$message` ?

stof commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -72,17 +80,29 @@ class Translator implements TranslatorInterface, TranslatorBagInterface
+            @trigger_error('Passing a MessageSelector instance into the '.__METHOD__.' as a second argument is deprecated since version 3.1 and will be removed in 4.0. Inject a MessageFormatterInterface[] instance instead.', E_USER_DEPRECATED);
+        }
+
+        if (null === $formatters || $formatters instanceof MessageSelector) {
```
I would handle the case of ``$formatters instanceof MessageSelector`` in the previous if, at the same time than triggering the deprecation. It will be more readable (we can have ``$formatters = array(new TransMessageFormatter());`` in the case of ``null`` which will be easier to understand what is passed IMO) and will need that we have a single block to remove in 4.0, and it is the one containing the deprecation trigger

stof commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -72,17 +80,29 @@ class Translator implements TranslatorInterface, TranslatorBagInterface
-     * @param string|null          $cacheDir The directory to use for the cache
-     * @param bool                 $debug    Use cache in debug mode ?
+     * @param string                      $locale     The locale
+     * @param MessageFormatterInterface[] $formatters An array of formatters
```
wrong phpdoc as it also supports ``MessageSelector|null``

stof replied:
and passing ``null`` is not deprecated

stof commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -50,9 +53,14 @@ class Translator implements TranslatorInterface, TranslatorBagInterface
+    /**
+     * @var string
+     */
+    private $defaultFormatter = TransMessageFormatter::NAME;
```
A better idea would be to take the name of the first formatter in the constructor

aitboudad committed:
[Translation] Support adding custom message formatter

aitboudad commented:
Hi, 
I removed the ability to use multiple formatters to simplify thing (it should be done in another PR).
I think this PR is ready now! ping @symfony/deciders

HeahDude commented on src/Symfony/Bundle/FrameworkBundle/Translation/Translator.php:
```diff
@@ -13,8 +13,9 @@
 use Symfony\Component\Translation\Translator as BaseTranslator;
-use Symfony\Component\Translation\MessageSelector;
 use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\Translation\MessageSelector;
```
IIUC This cs fix should not be targeting master to ease the merge of older branches

HeahDude commented on src/Symfony/Component/Translation/Formatter/PluralMessageFormatterInterface.php:
```diff
@@ -0,0 +1,30 @@
+     *
+     * @return string
+     */
+    public function pluralFormat($message, $number, $locale, array $parameters = array());
```
I'm not a native but what about `formatPlural`?

HeahDude commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -72,17 +75,25 @@ class Translator implements TranslatorInterface, TranslatorBagInterface
+
+        if ($formatter instanceof MessageSelector) {
+            $this->formatter = new TransMessageFormatter($formatter);
+            @trigger_error('Passing a MessageSelector instance into the '.__METHOD__.' as a second argument is deprecated since version 3.1 and will be removed in 4.0. Inject a MessageFormatterInterface[] instance instead.', E_USER_DEPRECATED);
```
3.2 :)

HeahDude replied:
What about:
```php
@trigger_error(sprintf('Passing a "%s" instance into the "%s" as a second argument is deprecated since version 3.2 and will be removed in 4.0. Inject a "%s" implementation instead.', MessageSelector::class, __METHOD__, MessageFormatterInterface::class), E_USER_DEPRECATED)
```
?

aitboudad replied:
comments addressed, thanks!

HeahDude reviewed (commented):
Left some minor comments, looks really good! Thanks!

HeahDude commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml:
```diff
@@ -31,6 +31,12 @@
 
+        <service id="translator.formatter" alias="translator.formatter.trans_message" />
+
+        <service id="translator.formatter.trans_message" class="Symfony\Component\Translation\Formatter\TransMessageFormatter" public="false">
```
The framework bundle will now need the [translation component from `~3.2`](https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/composer.json#L36) to make tests green.

aitboudad replied:
done

aitboudad commented:
Travis/fabbot failures is not related.

sstok commented on src/Symfony/Component/Translation/Formatter/TransMessageFormatter.php:
```diff
@@ -0,0 +1,48 @@
+/**
+ * @author Abdellatif Ait boudad <a.aitboudad@gmail.com>
+ */
+class TransMessageFormatter implements MessageFormatterInterface, PluralMessageFormatterInterface
```
`MessageFormatterInterface` is already implicitly included by `PluralMessageFormatterInterface`.

aitboudad replied:
fixed

aitboudad commented:
@symfony/deciders any chance to get your votes :) ?

javiereguiluz commented:
I still think the new proposal is a big step back in DX. Example:

#### Before

```php
$this->get('translator')->transChoice('...', 10, ['%count%' => 10]);
```

#### After

```php
use Symfony\Component\Translation\Formatter\TransMessageFormatter;

$this->get('translator')->trans('...', 10, [TransMessageFormatter::TRANS_CHOICE => 10]);
```

aitboudad commented:
@javiereguiluz please review the code, I abandoned the idea of multiple formatters at least right now, thanks!

javiereguiluz commented:
@aitboudad I'm sorry. Could it be possible to get a before/after comparison of the changes?

Here it's the before in PHP and twig:


```php
// simple
$this->get('translator')->trans('Hello');

// placeholders
$this->get('translator')->trans('Hello %name%', ['%name%' => 'Fabien']);

// pluralization
$this->get('translator')->transChoice(
    'There is one apple|There are %count% apples',
    10,
    array('%count%' => 10)
);

// selecting the catalogue
$this->get('translator')->trans('Hello', [], 'admin');

// selecting the locale
$this->get('translator')->trans('Hello', [], 'messages', 'fr_FR');
```

```twig
{# simple #}
{% trans %}Hello{% endtrans %}
{{ 'Hello'|trans }}

{# placeholders #}
{% trans with { '%name%': 'Fabien' } %}Hello %name%{% endtrans %}
{{ 'Hello'|trans({'%name%': 'Fabien'}) }}

{# pluralization #}
{% transchoice count %}
    {0} There are no apples|{1} There is one apple|]1,Inf[ There are %count% apples
{% endtranschoice %}

{{ '{0} There ...|{1} There ...'|transchoice(count) }}

{# selecting the catalogue #}
{% trans from 'admin' %}Hello{% endtrans %}
{{ 'Hello'|trans({}, 'app') }}

{# selecting the locale #}
{% trans into 'fr_FR' %}Hello{% endtrans %}
{{ 'Hello'|trans({}, 'messages', 'fr_FR') }}
```

aitboudad commented:
@javiereguiluz there are no changes on those parts :), the only thing to note is with this PR we can choose the way of formating translation message through config (Ex [intl formatter](https://github.com/symfony/symfony/pull/20007))
```yml
framework:
    translator:
        formatter: translator.formatter.intl
```

xabbuh commented on src/Symfony/Component/Translation/Formatter/TransMessageFormatter.php:
```diff
@@ -0,0 +1,48 @@
+/**
+ * @author Abdellatif Ait boudad <a.aitboudad@gmail.com>
+ */
+class TransMessageFormatter implements PluralMessageFormatterInterface
```
Not sure if this name is really good. Can we think about something more descriptive?

aitboudad replied:
what do you think about `NativeMessageFormatter` or `BasicMessageFormatter` ?

aitboudad replied:
or just `MessageFormatter` as it's the default one ?

xabbuh commented on src/Symfony/Component/Translation/Formatter/PluralMessageFormatterInterface.php:
```diff
@@ -0,0 +1,30 @@
+/**
+ * @author Abdellatif Ait boudad <a.aitboudad@gmail.com>
+ */
+interface PluralMessageFormatterInterface extends MessageFormatterInterface
```
Not sure if extending the `MessageFormatterInterface` makes sense here.

javiereguiluz replied:
Java API defines a `ChoiceFormat` message formatter ([see details](https://docs.oracle.com/javase/7/docs/api/java/text/ChoiceFormat.html#ChoiceFormat(java.lang.String))) Maybe it's a better alternative to `PluralMessageFormatter` ?

aitboudad replied:
LGTM

xabbuh commented:
You also need to document the deprecation in the upgrade files.

fabpot commented on UPGRADE-3.2.md:
```diff
@@ -147,3 +147,27 @@ Yaml
+Translator
+----------
+
+ * Passing a `Symfony\Component\Translation\MessageSelector` to `Translator` is not supported anymore.
```
Actually, this is still supported, but deprecated.

aitboudad replied:
fixed

aitboudad commented:
ping @fabpot

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml:
```diff
@@ -31,6 +31,12 @@
             <argument type="service" id="translator.selector" />
         </service>
 
+        <service id="translator.formatter" alias="translator.formatter.default" />
```
This is not needed (it's set anyway in the extension due to our default values).

aitboudad replied:
it's set only when the translator is enabled

xabbuh replied:
ah indeed, thanks

aitboudad commented on src/Symfony/Bundle/FrameworkBundle/Translation/Translator.php:
```diff
@@ -47,14 +48,15 @@ class Translator extends BaseTranslator implements WarmableInterface
+     * @param ContainerInterface                             $container A ContainerInterface instance
+     * @param MessageFormatterInterface|MessageSelector|null $formatter The message formatter
+     *                                                                  Passing the MessageSelector as a second parameter is deprecated since version 3.2
+     * @param array                                          $loaderIds An array of loader Ids
```
yep, It should be done in an another PR

xabbuh commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -74,17 +77,25 @@ class Translator implements TranslatorInterface, TranslatorBagInterface
+        if ($formatter instanceof MessageSelector) {
+            $this->formatter = new MessageFormatter($formatter);
+            @trigger_error(sprintf('Passing a "%s" instance into the "%s" as a second argument is deprecated since version 3.2 and will be removed in 4.0. Inject a "%s" implementation instead.', MessageSelector::class, __METHOD__, MessageFormatterInterface::class), E_USER_DEPRECATED);
+        } elseif (null === $formatter) {
```
We never set the `$formatter` property when `$formatter` is an instance of `MessageFormatterInterface`.

aitboudad replied:
fixed

xabbuh commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -192,17 +203,17 @@ public function trans($id, array $parameters = array(), $domain = null, $locale
-            '%count%' => $number,
-        ), $parameters);
+        if (!$this->formatter instanceof ChoiceMessageFormatterInterface) {
+            throw new InvalidArgumentException(sprintf('The formatter "%s" does not support the plural translations.', get_class($this->formatter)));
```
[...] does not support plural translations.

aitboudad replied:
fixed

xabbuh commented:
👍 

Status: Reviewed

Koc commented:
@aitboudad is it possible just use message formatter without translation inside Twig/PHP? For example when we need site with only one locale but need pluralization rules

aitboudad commented:
@Koc I guess yes and I think it's something nice to have but needs some additional work :), the only required part is (formatter + key/value store)

aitboudad commented:
@symfony/deciders Is there anything else before merging this PR ?

xabbuh commented:
Should be rebased. But otherwise  this looks good to me.

nicolas-grekas commented:
@aitboudad could you rebase again please?

aitboudad commented:
Rebased :), the requested changes from `fabbot.io` seems wrong

nicolas-grekas commented:
ping @symfony/deciders (and maybe @Nyholm) any reason not to merge?

nicolas-grekas commented:
Hum I might answer myself, because BC breaks?
@aitboudad can we add a BC layer and add deprecations to remove the BC break?
Should target 3.4 btw (with no BC break of course).

aitboudad commented:
@nicolas-grekas already done, there is no BC break.

Nyholm commented on src/Symfony/Component/Translation/Formatter/MessageFormatterInterface.php:
```diff
@@ -0,0 +1,30 @@
+     *
+     * @return string
+     */
+    public function format($message, $locale, array $parameters = array());
```
I fail to understand why we would need $locale here. We do not use it and I can't come up with any valid example were it is needed.

aitboudad replied:
see the use-case [IntlMessageFormatter](https://github.com/aitboudad/symfony/blob/7c3ac84d41570270b8cfcaae4c87a0b5c1384c86/src/Symfony/Component/Translation/Formatter/IntlMessageFormatter.php#L25) (https://github.com/symfony/symfony/pull/20007)

Nyholm replied:
Thank you

Nyholm reviewed (commented):
I think this looks good. It is another abstraction layer, but it makes sense. I only have one small comment.

HeahDude commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -925,6 +925,7 @@ private function registerTranslatorConfiguration(array $config, ContainerBuilder
 
         // Use the "real" translator instead of the identity default
         $container->setAlias('translator', 'translator.default');
+        $container->setAlias('translator.formatter', $config['formatter']);
```
What about:
```php
use Symfony\Component\Translation\Formatter\MessageFormatterInterface;
use Symfony\Component\Translation\Formatter\MessageFormatter;

//...
$container->setAlias(MessageFormatterInterface::class, $config['formatter']);
$container->setAlias(MessageFormatter::class, $config['formatter']);
```
?

nicolas-grekas replied:
:-1: for FQCN here, that'd be inconsistent
but the alias should be private, isn't it?
about the MessageFormatterInterface alias, that'd not for this PR since that'd be a "new feature".
but we did not add this alias in 3.3. Is there any real use case for type hinting MessageFormatterInterface and thus have it autowired?

nicolas-grekas replied:
@HeahDude got it, can you review again that part?

HeahDude commented on src/Symfony/Component/Translation/Formatter/MessageFormatter.php:
```diff
@@ -0,0 +1,48 @@
+     */
+    public function format($message, $locale, array $parameters = array())
+    {
+        return strtr((string) $message, $parameters);
```
Is this casting really needed? Passing anything else but a string would not respect the interface.

HeahDude commented on src/Symfony/Component/Translation/Formatter/MessageFormatter.php:
```diff
@@ -0,0 +1,48 @@
+    {
+        $parameters = array_merge(array('%count%' => $number), $parameters);
+
+        return $this->format($this->selector->choose($message, (int) $number, $locale), $locale, $parameters);
```
Same question for the int.

aitboudad replied:
I prefer to keep this one as it's used before and might break some apps if we remove it.

HeahDude commented on src/Symfony/Component/Translation/Translator.php:
```diff
@@ -192,17 +204,17 @@ public function trans($id, array $parameters = array(), $domain = null, $locale
-            '%count%' => $number,
-        ), $parameters);
+        if (!$this->formatter instanceof ChoiceMessageFormatterInterface) {
+            throw new InvalidArgumentException(sprintf('The formatter "%s" does not support plural translations.', get_class($this->formatter)));
```
The formatter is not an argument here, what about throwing a LogicException instead?

nicolas-grekas commented:
@aitboudad I changed the branch to 3.4, but I cannot push --force on your fork, so here is the rebase:
https://github.com/nicolas-grekas/symfony/tree/translation-formatters

aitboudad commented:
@nicolas-grekas you can push now!

nicolas-grekas commented:
@aitboudad thanks, I just push forced one last time, addressing a few minor things meanwhile.

HeahDude commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml:
```diff
@@ -28,6 +28,13 @@
         </service>
 
+        <service id="translator.formatter" alias="translator.formatter.default" />
+        <service id="Symfony\Component\Translation\Formatter\MessageFormatterInterface" alias="formatter" />
```
`alias="translator.formatter"`?

nicolas-grekas replied:
fixed thanks

nicolas-grekas commented:
I think this PR is ready to merge, failures are false positives.
@aitboudad would you be able to work on a doc PR if worth it?

aitboudad commented:
why not :), I'll try to work on it next weekend.

aitboudad commented:
@nicolas-grekas I missed to mention the doc PR #8284, I guess the PR is ready to merge.

nicolas-grekas reviewed (approved):
@xabbuh I'll let you merge if you don't mind

xabbuh commented on UPGRADE-3.4.md:
```diff
@@ -234,6 +234,27 @@ Translation
+ * Passing a `Symfony\Component\Translation\MessageSelector` to `Translator` has been
+   deprecated. You should pass a message formatter instead
+
+  Before:
```
Just noticed: This line and the following need to be indented by an additional space so that it will be rendered property as part of the list item.

aitboudad replied:
fixed

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/translation.xml:
```diff
@@ -28,6 +28,12 @@
             <tag name="monolog.logger" channel="translation" />
         </service>
 
+        <service id="translator.formatter" alias="translator.formatter.default" />
```
Do we need this? I mean this is just redundant with what we do in the extension (in combination with the default option value).

xabbuh replied:
Oh, in fact there is one difference: The alias registered in the extension is private. So IMO that's a good argument to remove this one here. At least it will avoid confusion when reading the code.

nicolas-grekas replied:
(this one also is private, all file start with public="false" as default ;) )

xabbuh replied:
hm okay 😆

aitboudad replied:
the default value can be overridden throughout the config, without using the alias we can't inject the default one if we want.

xabbuh replied:
But we always create the alias in https://github.com/aitboudad/symfony/blob/fd9f484879a08e00a42a3888d8b1681923b8eb0b/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L1078, don't we (`$config['formatter']` will have the default value if the user didn't provide any)?

aitboudad replied:
you're right, removed!

xabbuh commented on UPGRADE-3.4.md:
```diff
@@ -234,6 +234,27 @@ Translation
    and will be removed in 4.0, use `Symfony\Component\Translation\Writer\TranslationWriter::write` 
    instead. 
 
+ * Passing a `Symfony\Component\Translation\MessageSelector` to `Translator` has been
```
Should probably be added to the component's changelog file too. And we also need to update the `UPGRADE-4.0.md` file accordingly.

aitboudad replied:
done

xabbuh commented:
Thank you @aitboudad.
